### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by Shevtsooov**
[DEMO LINK](https://Shevtsooov.github.io/react_todo-app-loading-todos/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -14,143 +60,31 @@ export const App: React.FC = () => {
...
+
+        {filteredTodos && (
+          <Footer
+            visibleTodos={visibleTodos}
+            status={status}
+            setStatus={setStatus}
+            areCompleted={areCompleted}
+            handleClearCompleted={handleClearCompleted}
+          />
+        )}
```
Comment (maksym-bobryk) [inline src/App.tsx:79]: According to the task you need to hide a list of todos and a footer if there aren't any todos

File: src/components/Footer/Footer.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  areCompleted,
+  handleClearCompleted,
+}) => {
+  return (
+    <footer className="todoapp__footer">
+      <span className="todo-count">
+        {visibleTodos?.length !== 1
+          ? `${visibleTodos?.length} items left`
+          : `${visibleTodos.length} item left`}
+      </span>
```
Comment (maksym-bobryk) [inline src/components/Footer/Footer.tsx:26]: Nice one!

File: src/components/Todolist/Todolist.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+          >
+            ×
+          </button>
+
+          <div className="modal overlay">
+            <div className="modal-background has-background-white-ter" />
+            <div className="loader" />
+          </div>
+        </div>
+      ))}
```
Comment (maksym-bobryk) [inline src/components/Todolist/Todolist.tsx:39]: Haven't added a key and maybe it would be better to make a separate component for todo info

**Review by maksym-bobryk — CHANGES_REQUESTED** [review]
Almost done but good job🚀

**Comment by Shevtsooov** [comment]
@maksym-bobryk Hey, thanks for the points.
All of them have been fixed)

File: src/App.tsx [hunk 4]
```diff
@@ -1,10 +1,59 @@
...
+import { Todolist } from './components/Todolist/Todolist';
+import { ErrorMessage } from './types/ErrorMessage';
+import { TodoStatus } from './types/TodoStatus';
 
-const USER_ID = 0;
+const USER_ID = 10890;
 
 export const App: React.FC = () => {
+  const [visibleTodos, setVisibleTodos] = useState<Todo[]>([]);
+  const [status, setStatus] = useState<TodoStatus>(TodoStatus.ALL);
```
Comment (silchyn) [inline src/App.tsx:17]:
```suggestion
  const [filter, setFilter] = useState<TodoStatus>(TodoStatus.ALL);
```

File: src/App.tsx [hunk 5]
```diff
@@ -1,10 +1,59 @@
...
+      .catch(() => {
+        setIsError(ErrorMessage.LOADERROR);
+      });
+  }, []);
+
+  const handleCloseError = () => {
+    setIsError(ErrorMessage.NOERROR);
+  };
+
+  const filterTodos = useCallback((todoStatus: string) => {
```
Comment (silchyn) [inline src/App.tsx:32]:
```suggestion
  const filterTodos = useCallback((filter: TodoStatus) => {
```

File: src/App.tsx [hunk 6]
```diff
@@ -1,10 +1,59 @@
...
+    if (todoStatus === TodoStatus.COMPLETED) {
+      return visibleTodos.filter(todo => todo.completed);
+    }
+
+    if (todoStatus === TodoStatus.ACTIVE) {
+      return visibleTodos.filter(todo => !todo.completed);
+    }
+
+    return visibleTodos;
```
Comment (silchyn) [inline src/App.tsx:41]: Wouldn't `switch-case` fit here?

File: src/App.tsx [hunk 7]
```diff
@@ -1,10 +1,59 @@
...
+  }, [status, visibleTodos]);
+
+  const filteredTodos = filterTodos(status);
```
Comment (silchyn) [inline src/App.tsx:44]: We don't use the function anywhere else; can't we calculate `filteredTodos` here?

File: src/App.tsx [hunk 8]
```diff
@@ -1,10 +1,59 @@
...
+
+  const areCompleted = filteredTodos.filter(todo => todo.completed);
+  const areActive = filteredTodos.filter(todo => !todo.completed);
```
Comment (silchyn) [inline src/App.tsx:47]:
```suggestion
  const completedTodos = filteredTodos.filter(todo => todo.completed);
  const activeTodos = filteredTodos.filter(todo => !todo.completed);
```

File: src/App.tsx [hunk 9]
```diff
@@ -1,10 +1,59 @@
...
+
+  const handleClearCompleted = () => (
+    filteredTodos.filter(todo => !todo.completed)
+  );
```
Comment (silchyn) [inline src/App.tsx:51]: What's the point?

File: src/components/Footer/Footer.tsx [hunk 10]
```diff
@@ -0,0 +1,71 @@
+import cn from 'classnames';
+import { Todo } from '../../types/Todo';
+import { TodoStatus } from '../../types/TodoStatus';
+
+type Props = {
+  visibleTodos: Todo[] | null,
+  status: string,
```
Comment (silchyn) [inline src/components/Footer/Footer.tsx:7]: But why not `TodoStatus`?

File: src/components/Footer/Footer.tsx [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+};
+
+export const Footer: React.FC<Props> = ({
+  visibleTodos,
+  status,
+  setStatus,
+  areCompleted,
+  handleClearCompleted,
+}) => {
+  return (
```
Comment (silchyn) [inline src/components/Footer/Footer.tsx:20]:
```suggestion
}) => (
```

File: src/components/Footer/Footer.tsx [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    <footer className="todoapp__footer">
+      <span className="todo-count">
+        {visibleTodos?.length !== 1
+          ? `${visibleTodos?.length} items left`
+          : `${visibleTodos.length} item left`}
```
Comment (silchyn) [inline src/components/Footer/Footer.tsx:25]:
```suggestion
        {visibleTodos?.length === 1
          ? `${visibleTodos.length} item left`
          : `${visibleTodos?.length} items left`}
```

File: src/components/Footer/Footer.tsx [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  status,
+  setStatus,
+  areCompleted,
+  handleClearCompleted,
+}) => {
+  return (
+    <footer className="todoapp__footer">
+      <span className="todo-count">
+        {visibleTodos?.length !== 1
+          ? `${visibleTodos?.length} items left`
```
Comment (silchyn) [inline src/components/Footer/Footer.tsx:24]: So what if `visibleTodos` is `null`?

File: src/components/Footer/Footer.tsx [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+          className={cn('filter__link', {
+            selected: status === TodoStatus.COMPLETED,
+          })}
+          onClick={() => setStatus(TodoStatus.COMPLETED)}
+        >
+          Completed
+        </a>
+      </nav>
+
+      {!areCompleted && (
```
Comment (silchyn) [inline src/components/Footer/Footer.tsx:60]: Who are completed?

File: src/components/TodoInfo/TodoInfo.tsx [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        ×
+      </button>
+
+      <div className="modal overlay">
+        <div className="modal-background has-background-white-ter" />
+        <div className="loader" />
+      </div>
+    </div>
+  );
+};
```
Comment (silchyn) [inline src/components/TodoInfo/TodoInfo.tsx:38]:
```suggestion
export const TodoInfo: React.FC<Props> = ({ todo }) => (
  <div
    className={cn('todo', {
      completed: todo.completed,
    })}
  >
    <label className="todo__status-label">
      <input
        type="checkbox"
        className="todo__status"
        checked
      />
    </label>

    <span className="todo__title">{todo.title}</span>

    <button
      type="button"
      className="todo__remove"
    >
      ×
    </button>

    <div className="modal overlay">
      <div className="modal-background has-background-white-ter" />
      <div className="loader" />
    </div>
  </div>
);
```

File: src/components/Todolist/Todolist.tsx [hunk 16]
```diff
@@ -0,0 +1,16 @@
+import { Todo } from '../../types/Todo';
+import { TodoInfo } from '../TodoInfo/TodoInfo';
+
+type Props = {
+  filteredTodos: Todo[] | null;
```
Comment (silchyn) [inline src/components/Todolist/Todolist.tsx:5]: Should it accept `null`?

File: src/types/ErrorMessage.ts [hunk 17]
```diff
@@ -0,0 +1,7 @@
+export enum ErrorMessage {
+  NOERROR = '',
+  LOADERROR = 'Unable to load a todo',
+  ADDERROR = 'Unable to add a todo',
+  DELETEERROR = 'Unable to delete a todo',
+  UPDATEERROR = 'Unable to update a todo',
```
Comment (silchyn) [inline src/types/ErrorMessage.ts:6]: Should it be uppercase?

**Review by silchyn — CHANGES_REQUESTED** [review]

File: src/components/Footer/Footer.tsx (hunk 13 above)
Comment (Shevtsooov) [reply src/components/Footer/Footer.tsx]: @silchyn Should I create another condition here?

**Comment by Shevtsooov** [comment]
@silchyn Thanks, I have fixed most of the issues, still there is one I am not sure about.
Please check my comment above.

File: src/components/Footer/Footer.tsx (hunk 13 above)
Comment (silchyn) [reply src/components/Footer/Footer.tsx]: That depends on the logic you're building. Should `Footer` be rendered if `visibleTodos === null`? If yes, what should the user see?

File: src/App.tsx [hunk 18]
```diff
@@ -1,156 +1,89 @@
...
-            {/* This form is shown instead of the title and remove button */}
-            <form>
-              <input
-                type="text"
-                className="todo__title-field"
-                placeholder="Empty todo will be deleted"
-                value="Todo is being edited now"
-              />
-            </form>
+  const filteredTodos = filterTodos();
```
Comment (silchyn) [inline src/App.tsx:45]: > We don't use the function anywhere else; can't we calculate `filteredTodos` here?

File: src/components/Footer/Footer.tsx [hunk 19]
```diff
@@ -0,0 +1,66 @@
+import cn from 'classnames';
+import { Todo } from '../../types/Todo';
+import { TodoStatus } from '../../types/TodoStatus';
+
+type Props = {
+  visibleTodos: Todo[] | null,
+  filter: string,
+  setFilter: (status: TodoStatus) => void,
+  areCompleted: Todo[] | null,
```
Comment (silchyn) [inline src/components/Footer/Footer.tsx:9]: Who are completed, and should the array be named like that?

File: src/components/Header/Header.tsx [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+      <button
+        type="button"
+        className={cn('todoapp__toggle-all', {
+          active: areActive,
+        })}
+        aria-label="button"
+      />
+
+      <form
+        onSubmit={(e) => handleSubmit(e)}
```
Comment (silchyn) [inline src/components/Header/Header.tsx:25]:
```suggestion
        onSubmit={handleSubmit}
```

File: src/components/TodoInfo/TodoInfo.tsx [hunk 21]
```diff
@@ -0,0 +1,36 @@
+import cn from 'classnames';
+import { Todo } from '../../types/Todo';
+
+type Props = {
+  todo: Todo
+};
+
+export const TodoInfo: React.FC<Props> = ({ todo }) => (
```
Comment (silchyn) [inline src/components/TodoInfo/TodoInfo.tsx:8]: Destructure `todo`

File: src/components/Todolist/Todolist.tsx [hunk 22]
```diff
@@ -0,0 +1,16 @@
+import { Todo } from '../../types/Todo';
+import { TodoInfo } from '../TodoInfo/TodoInfo';
+
+type Props = {
+  filteredTodos: Todo[];
+};
+
+export const Todolist: React.FC<Props> = ({ filteredTodos }) => {
+  return (
```
Comment (silchyn) [inline src/components/Todolist/Todolist.tsx:9]:
```suggestion
export const Todolist: React.FC<Props> = ({ filteredTodos }) => (
```

File: src/types/TodoStatus.ts [hunk 23]
```diff
@@ -0,0 +1,5 @@
+export enum TodoStatus {
+  ALL = 'all',
+  ACTIVE = 'active',
+  COMPLETED = 'completed',
```
Comment (silchyn) [inline src/types/TodoStatus.ts:4]: Should it be uppercase?

**Review by silchyn — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 24]
```diff
@@ -1,10 +1,54 @@
...
+import { ErrorMessage } from './types/ErrorMessage';
+import { TodoStatus } from './types/TodoStatus';
 
-const USER_ID = 0;
+const USER_ID = 10890;
 
 export const App: React.FC = () => {
+  const [visibleTodos, setVisibleTodos] = useState<Todo[]>([]);
+  const [filter, setFilter] = useState<TodoStatus>(TodoStatus.all);
+  const [isError, setIsError] = useState<ErrorMessage>(ErrorMessage.noError);
```
Comment (silchyn) [inline src/App.tsx:18]: Name `isError` supposes it holds a `boolean` value

File: src/App.tsx [hunk 25]
```diff
@@ -1,10 +1,54 @@
...
+      .catch(() => {
+        setIsError(ErrorMessage.loadError);
+      });
+  }, []);
+
+  const handleCloseError = () => {
+    setIsError(ErrorMessage.noError);
+  };
+
+  const filteredTodos = useCallback(() => {
```
Comment (silchyn) [inline src/App.tsx:32]: Should function be named like that?

Comment (silchyn) [reply src/App.tsx:32]: And should it be a function?

**Review by silchyn — APPROVED** [review]

**Pull request closed without merging**